### ISSUE TITLE
Apparent software bottleneck from Racket background expansion in DrRacket v7.6 [cs] on Debian 10

### ISSUE THREAD

**Issue by MartianMoorGrove** [racket/drracket#350]
**Description:** 
    Racket language background expansion in DrRacket v7.6 [cs] causes the entire DrRacket program to severely hang or lag on anything and everything I type into the definitions window, making that window nearly unusable. This problem is not due to any particular module and does not affect the rest of my computer. Raising the memory limit doesn't matter and I don't see any hardware constraints in my system resource monitor. I've had no problems using previous versions of DrRacket [cs]. Disabling background expansion in Preferences is a quick-fix, and I can use the "Check Syntax" button to manually get the same functionality that I'd want.

**Steps to Reproduce:**
    Using the official racket-7.6-x86_64-linux-cs.sh installer, install Racket v7.6 [cs] into your home directory as ~/racket. Start DrRacket from the command line. Make sure background expansion is enabled and that the Racket language is selected. Start typing anything into the definitions window.

**System info:**  
-   OS: Debian GNU/Linux 10 x86-64
-   Kernel version: 1 SMP Debian 4.19.98-1 (2020-01-26)
-   Processor: Intel® Core™ i7 CPU Q 720 @ 1.60GHz × 8 cores
-   RAM: 5.8GiB usable

**Comment by MartianMoorGrove** [comment]
This problem still exists in DrRacket v7.7 [cs]. DexterLagan may be referring to the same thing in their newer post at [https://github.com/racket/drracket/issues/376](url) .

**Comment by MartianMoorGrove** [comment]
@rfindler 
Should I close this issue? I'm using a relatively new workstation now so my context is different, but with it DrRacket v7.9 [cs] runs fine so long as the definitions window isn't larger than 100-characters by 50-lines in size. Any larger and I still have to disable background expansion for things to be smooth, though I'm not sure anyone needs it larger than that.

Issue #422 covers the same problem and is on the first issue page. And it's 20 times less verbose lol. :)

**Comment by rfindler** [comment]
I definitely know that there are problems with DrRacket being slow to refresh when the window size gets large. It may be interesting to track down exactly why DrRacket is trying to refresh in your specific situation, however. It should be something that's related to something you're doing to DrRacket. I mean, if you just sit there and don't do anything or if you make very simple edits (andding and removing spaces with the mouse not on top of the definitions window) then I guess you won't see any particularly bad behavior (or if you do that's interesting!). There is probably some specific series of operations you're doing that's making DrRacket decide it needs to do a lot of work and if we knew what those operations you are doing and then we might be able to figure out why DrRacket is deciding to do a lot of work and then we might be able to fix something.

But without those steps it does seem better to just close the issue. I don't really have a toe hold here currently :(

**Comment by MartianMoorGrove** [comment]
@rfindler 
I should have gone back and reassessed the issue now that I'm more experienced; lesson learned. Having now done so, I found that the hanging/lagging occurs while there's a known error somewhere in the definitions window. That is basically until most expressions or forms are finished. Writing a long expression with several layers of nested parentheses becomes problematic.

The problem is avoided by some ratio of window-size to words-per-minute typed. At larger than 100-characters by 50-lines I have to type at over 60-80 wpm (using the five character standard) for that to happen, which is faster than the average person can type.

The problem does not appear in DrRacket v7.5 [cs], even with the definitions window completely maximized to 249-characters by 62-lines. However, if I run the v7.5 drracket.rkt source file from the command line using Racket v7.9 [cs], the problem reappears. This implies to me the "possibility" that DrRacket is not the source of the problem.

The problem started in the v7.6 [cs] distribution.

My system is still Debian GNU/Linux 10 x86-64.
Hopefully I've given enough information, and I'll make sure I'm better at this stuff in the future.

**Comment by rfindler** [comment]
@MartianMoorGrove DrRacket will, for the purposes of uncovering a syntax error in your program, make a copy of the text of the buffer syncronously which means it will ignore keystrokes while it is doing that, which will make it feel laggy. I wonder if that's what you're seeing here. If so, it would be the [`fetch-data-to-send`](https://github.com/racket/drracket/blob/master/drracket/drracket/private/module-language.rkt#L1215) method, which is (called here)[https://github.com/racket/drracket/blob/master/drracket/drracket/private/module-language.rkt#L2279]. If you were to stick a printf there, maybe with a count of the time it takes to compute things, eg:

```
(define now (current-process-millisconds))
....
(printf "took ~a msec\n" (- (current-process-milliseconds) now))
```

and those times were longish in situations that you're seeing laggy behavior and not in others then that's probably the culprit.

**Comment by MartianMoorGrove** [comment]
@rfindler I changed that area of the code to look like this:
```
   (when dirty/pending-tab
      (define now (current-process-milliseconds))
      (define-values (editor-contents filename/loc) 
        (send (send dirty/pending-tab get-defs) fetch-data-to-send))
      (printf "took ~a msec\n" (- (current-process-milliseconds) now))
```
With an error in a maximized definitions window, it prints "took 0 msec" no matter what I type in. I assume that means the time taken is some fraction of 1 msec which would mean this method isn't the culprit as the lagging with the window maximized appears to around 900 msec. Again, if I type rather fast the culprit is avoided/delayed somehow. That's not required in v7.5, I can type as slow or fast as I want even with the window maximized.

I actually discovered that method call a few months ago and wrapped it in a `time` form and received the same results. I just wanted to make sure the results were in fact the same.

**Comment by rfindler** [comment]
Okay, guess that wasn't it then. There are other things that are triggered with some delay based on typing (so typing quickly would cause them to never happen as each keystroke would push them out further). Do you have the search window open (menukey-f)? How about the contour (menukey-u)?

**Comment by MartianMoorGrove** [comment]
The search window isn't open and the contour isn't shown.

The culprit appears to be within this line of logic starting on this [line](https://github.com/racket/drracket/blob/8f4af031be9fe218ca17366ffb9369302b75cbee/drracket/drracket/private/module-language.rkt#L2384):
`oc-timer-expired`->`send-to-place`->`show-results`(which is a reference to oc-finished)->2nd usage of `set-oc-status`->`update-gui`->`update-error-in-defs`->`end-edit-sequence`->`after-edit-sequence`->`oc-set-dirty`->2nd usage of `set-oc-status`->`update-gui`->`update-error-in-defs`->`end-edit-sequence`.

Am I correct in assuming that `after-edit-sequence` is called in response to `end-edit-sequence`?

When I wrap a `time` form around of the bodies of `oc-finished` and `oc-set-dirty` I see a combined lag of roughly 875 msec, which when added together is close to that 900 msec estimate that I mentioned. 

I'll look around more sometime over the weekend.

**Comment by rfindler** [comment]
That is very interesting @MartianMoorGrove ! I don't see that on my machine when I wrap the same things in `time`. Since that's enough time that we might get some useful information, can you share the output you see when you wrap them in `profile` (or use `profile-thunk`)?  The library is also named just `profile` so you'd need to add `(require profile)` to the top of the file.

And yes, `end-edit-sequence` calls `after-edit-sequence` (this is set up for callbacks that want to do work when, say, an insert happens, can also track if the insertion is happening during an edit sequence and postpone the work until the end of the edit sequence).

**Comment by MartianMoorGrove** [comment]
@rfindler You were able to see my original unedited comment? I was extremely unconfident in what I said so I edited it out and eventually deleted that edit. I chose `oc-finished` and `oc-set-dirty` because they share a call to `set-oc-status` which eventually reaches `end-edit-sequence` which makes up the bulk of the lag.

Here are the outputs:
- From the oc-finished body:
  
  `Profiling results`
  `-----------------`
  `Total cpu time observed: 0ms (out of 432ms)
   Number of samples taken: 1 (once every 0ms)`

- From the oc-set-dirty body:
  
  `Profiling results`
  `-----------------`
  `Total cpu time observed: 0ms (out of 415ms)
   Number of samples taken: 1 (once every 0ms)`

 
 For some reason `Caller`, `Indx`, `Total ms (pct)` and the rest are blank in both outputs. Maybe it has something to do with `oc-finished` being called by a different thread and `oc-set-dirty` in the augmented `after-edit-sequence` possibly being on a different thread too(?).

An interesting output from `end-edit-sequence` on [this](https://github.com/racket/drracket/blob/8f4af031be9fe218ca17366ffb9369302b75cbee/drracket/drracket/private/module-language.rkt#L1113) line:

  `profiled thunk error: sequence-contract-violation: negative: method end-edit-sequence cannot be called, except in states (in-edit-sequence)`
`Profiling results`
`-----------------`
  `Total cpu time observed: 0ms (out of 418ms)
  Number of samples taken: 0 (once every 0ms)`

**Comment by MartianMoorGrove** [comment]
By the way, when on [this](https://github.com/racket/drracket/blob/8f4af031be9fe218ca17366ffb9369302b75cbee/drracket/drracket/private/module-language.rkt#L2387) line I change 250 to 1000 msec (exceeding the lag time) the problem is largely mitigated, although error-catching suffers for it. I'm guessing it helps because the edit sequences have time to complete.

**Comment by rfindler** [comment]
Some answers to specific questions: I think I didn't see the original version; I was certainly replying to the version I see there now. All of the edit sequences (and other things that modify the buffer's state or that you'd perceive as lag) happen on the same thread. Looking at the code, `oc-finished` is happening on that same thread as is `oc-set-dirty`. 

Bummer about the profiling results.

I looked through the work that gets done after an edit sequence but am not seeing anything obvious. The code is mostly [here](https://github.com/racket/gui/tree/master/gui-lib/framework/private) and [here](https://github.com/racket/drracket/tree/master/drracket/drracket/private) but I'm not seeing anything obvious that would take remotely that long.

One thing I should have asked earlier: could you try a more recent (perhaps git-based) build? It should be easy to build and the standard way to build is just to clone the repo and do "make" at the top-level (the top-level makefile does all the configuring and whatnot that's required). This will produce a version of racket (and drracket) that you run right out of that directory and can just be deleted if you don't want it. It should build without any fuss if you have the compiler and whatnot installed.

**Comment by MartianMoorGrove** [comment]
@rfindler 
Yep, I had lost track; they are all on that same thread.

I cloned the repository and ran `make` successfully. The issue/problem does not exist in DrRacket version 8.0.0.1--2021-01-10(-/f) [cs]. Having an error somewhere in the definitions window does not cause any lagging upon keystrokes. Everything seems good, though I've no clue why.

**Comment by rfindler** [comment]
It may be that we've fixed the bug along the way. 7.6 was a while ago and lots of improvements inside CS have happened since then.

**Comment by MartianMoorGrove** [comment]
@rfindler I was using v7.9 for this discussion; the issue started in 7.6.

**Comment by rfindler** [comment]
Oh, yes. You did say that. Sorry!

**Comment by MartianMoorGrove** [comment]
Is there anything else I can do? I will close this issue otherwise.

**Comment by rfindler** [comment]
Well, I'm not sure I'm really getting what's going on, actually. How would you feel to make a screen recording showing the bad behavior? I'm not saying you need to narrate it or anything like that, just do what you're doing and then tell me where the bad behavior shows up in the video?

**Comment by MartianMoorGrove** [comment]
Sure thing. The video is [here](https://youtu.be/MDKsmKnBexk); be sure to increase the definition if it's low. The overhead of the screen recording multiplies the lag by maybe 1.5, roughly. My cursor has no perceivable impact whether on or off the window.

I'm typing at roughly 40 wpm here, pausing a few seconds after arriving at my next code line so that DrRacket catches up.
I start by entering a `struct` form. The lagging starts with the first few characters. I purposely make an error with `strct` so as to keep an error in in the window as I move to the next four lines of code. The lagging continues cumulatively from each keystroke as I create those four `struct` instances. Near the end I copy that long `struct` instance, correct `strct` up in the form, and paste that long instance several times. As there is no error in the `struct` form and the pasted instances are syntactically correct, there is no lag. If I paste them with an incorrect constructor such as `b`, the pastes lag. Likewise if I were to type them in, they would each would be a syntax error until completed and cause lagging.

**Comment by rfindler** [comment]
Okay, wow. That's really bad! I'll think about what question to ask next but I don't have a good idea at the moment.

**Comment by MartianMoorGrove** [comment]
@rfindler Okay. Let me know if I can be of any further help; this has been a good learning experience for me. I'll keep on learning more in the mean time. :)

**Comment by MartianMoorGrove** [comment]
@rfindler I had an idea:
In my third comment I mentioned that I ran DrRacket's v7.5 [cs] drracket.rkt module (I did equally run main.rkt) using Racket v7.9 [cs], and the problem appeared. Now I reversed that, and ran DrRacket's v7.9 [cs] main.rkt module using Racket v7.5 [cs] and everything works fine; no lagging. Leaving out its dependencies, does that rule out the DrRacket package as having contained the issue?

**Comment by MartianMoorGrove** [comment]
@rfindler 
Inspired by my previous comment's mistake, I found out where the problem is. It's in at least the racket-7.9-x86_64-linux-cs.sh installer. Since this issue started with v7.6 [cs] I'll assume it's in that installer, up to and including the v7.9 [cs] one and I'm worried it will end up in the upcoming v8.0 [cs] one.

To ensure a clean reference, I reinstalled Racket with racket-7.9-x86_64-linux-cs.sh after redownloading it from [here](https://download.racket-lang.org) and verifying its checksum. Then I ran DrRacket and experienced the usual lagging problem we've discussed.

After that, to ensure a clean build, I ran `raco setup -c` to delete the precompiled files from the installer, and ran `raco setup`. DrRacket had no lag after that.

I repeated this process three times to confirm.

**Comment by rfindler** [comment]
Oh interesting! That could mean that zo file dates are somehow out of sync in the installer maybe? And so DrRacket is loading a lot of code somehow when creating a new namespace for evaluating your program?

Do you see the same problem with the [release candidate](https://pre-release.racket-lang.org/)?

And also perhaps don't bother with the `-c` step. Just run `raco setup` and see if that fixes the problem?

**Comment by rfindler** [comment]
@MartianMoorGrove actually, run `raco setup -v` and we'll be able to see if the dates are wrong in the bundle.

**Comment by MartianMoorGrove** [comment]
Yes, the problem is also in the release candidate.

I reinstalled Racket with racket-7.9-x86_64-linux-cs.sh, and ran `raco setup -v`. There were no errors reported and the problem still existed. I did the same with the release candidate and had the same results.

Just to be clear, [here](https://lh3.googleusercontent.com/meECpXmRw6PsE0006tC5kLNWpmECddKWRlCbX1TisIA_AJJHtVGTnYCTDHXQlPK2qnibYi6MT5zb2mcIWaqnKIyhYbeGnzNpCIh1NFzzgMqR9HOUd1nAbU1H07cVac-55EnxggGh=w1226-h754-no) is a screenshot of my system information output.

**Comment by MartianMoorGrove** [comment]
I never thought to test anything but DrRacket [cs] since it's what I've used and it's been the subject of this issue, but now having done so I find that the lagging problem also exists in DrRacket [3m], but to a much lesser extent. It's almost bearable. Likewise using `raco setup -c` and then `raco setup` fixes it.

**Comment by mflatt** [comment]
Thanks for your persistence here! I'm able to replicate the problem and see it go away with a recompile. Specifically, recompiling the file "bitmap.rkt" in the "draw-lib" package makes a big difference.

So far, it looks like it's an issue with going through an intermediate "machine independent" format. That's how distributions are built: packages are first complied to a form that loads slowly but can work with BC or CS on any platform, and then that form is recompiled to BC format (bytecode) or a specific platform's CS format (machine code).

If I dig out the machine-independent "bitmap_rkt.zo" and drop it in place, DrRacket is still slow. If I recompile from that "bitmap_rkt.zo", still slow. If I remove that ".zo" and recompile from source, DrRacket is much faster.

Going through machine-independent form isn't supposed to matter at all, but it sure seems like it does, so I will be investigating more.

**Comment by MartianMoorGrove** [comment]
@mflatt 

I'm glad to have been helpful. I learned quite a bit while helping @rfindler; thank you Robby for prompting me along.
Let me know if I can be of any further help and if I think of something I'll comment.

**Comment by hidaris** [comment]
@mflatt Can we include this fix in the upcoming 8.0?

**Comment by mflatt** [comment]
If it solves the problem in the next round of snapshots (new ones will be at the Utah site in about 12 hours), then I'll recommend it for v8.0.

**Comment by mflatt** [comment]
A snapshot is ready at https://www.cs.utah.edu/plt/snapshots/current/installers/racket-current-x86_64-linux-precise.sh , and it runs correctly after installing in my environment. Can you give that one a try, @MartianMoorGrove?

**Comment by MartianMoorGrove** [comment]
@mflatt 

DrRacket version 8.0.0.1--2021-01-13(c64bf5d/a) [cs] from the racket-current-x86_64-linux-precise.sh installer has no lagging. It appears to running correctly except for one error output: `Gdk: gdk_wayland_window_handle_configure_popup: assertion 'impl->transient_for' failed`. I don't know if that matters for the purpose of this snapshot.

**Comment by samth** [comment]
@MartianMoorGrove are you running under Wayland? (I don't think many of us do that, so that would explain why we hadn't seen that before.)

**Comment by MartianMoorGrove** [comment]
@samth Yes I am. From within the output of `env`: XDG_SESSION_TYPE=wayland .
I also ran "`loginctl show-session $(loginctl | grep $(whoami) | awk '{print $1}') -p Type`" just to double check and had an equivalent result.

**Comment by MartianMoorGrove** [comment]
@samth it also outputs `(drracket:5347): Gdk-WARNING **: 14:07:42.743: losing last reference to undestroyed window` on exiting the program. I wasn't sure if the error was in DrRacket so I didn't say anything. It also exists in v7.9 [cs] and possibly the previous ones; my apologies for being late with it.

**Comment by MartianMoorGrove** [comment]
The `undestroyed window` warning exists before and after the clean recompilation fix we've discussed in this issue.

**Comment by MartianMoorGrove** [comment]
This issue has been fixed with Racket v8.0.
